File: practice/2-search-insert-position/solution.go
package main

import "fmt"

type Solution1 struct{}

func (s Solution1) searchInsert(nums []int, target int) int {
	left, right := 0, len(nums)-1

	for left <= right {
		middle := (left + right) / 2
		if nums[middle] < target {
			left = middle + 1
		} else if nums[middle] > target {
			right = middle - 1
		} else {
			return middle
		}
	}

	return right + 1
}

type Solution2 struct{}

func (s Solution2) searchInsert(nums []int, target int) int {
	left, right := 0, len(nums)

	for left < right {
		middle := (left + right) / 2
		if nums[middle] < target {
			left = middle + 1
		} else if nums[middle] > target {
			right = middle
		} else {
			return middle
		}
	}

	return left
}

func main() {
	s1 := Solution1{}
	sr := s1.searchInsert([]int{1, 3, 5, 6}, 5)
	fmt.Println(sr)

	s2 := Solution2{}
	sc := s2.searchInsert([]int{1, 3, 5, 6}, 5)
	fmt.Println(sc)
}

File: practice/2-search-insert-position/solution.py
from typing import List


class Solution1:
  """解法一: 左闭右闭
  """
  def searchInsert(self, nums: List[int], target: int) -> int:
    left, right = 0, len(nums) - 1

    while left <= right:
      middle = left + (right - left) // 2
      if nums[middle] < target:
        left = middle + 1
      elif nums[middle] > target:
        right = middle - 1
      else:
        return middle
    
    return right + 1
  

class Solution2:
  """解法二: 左闭右开
  """
  def searchInsert(self, nums: List[int], target: int) -> int:
    left, right = 0, len(nums)

    while left < right:
      middle = left + (right - left) // 2
      if nums[middle] < target:
        left = middle + 1
      elif nums[middle] > target:
        right = middle
      else:
        return middle
      
    return left